master/static/migrations/20240508134054_add-run-hparam-table.tx.up.sql

```diff
@@ -0,0 +1,42 @@
+CREATE TABLE run_hparams (
+    run_id int REFERENCES runs(id) ON DELETE CASCADE,
+    hparam text NOT NULL,
+    number_val float NULL,
+    text_val text NULL,
+    bool_val boolean NULL
+);
+
+INSERT INTO run_hparams(run_id, hparam, number_val, text_val, bool_val)
+SELECT s.id as run_id, s.key as hparam,
+CASE WHEN s.type='number' THEN s.value::text::float ELSE NULL END as number_val,
+CASE WHEN s.type='string' THEN s.value::text ELSE NULL END as text_val,
+CASE WHEN s.type='boolean' THEN s.value::text::boolean ELSE NULL END as bool_val
+FROM (SELECT r.id, h.key, h.value, jsonb_typeof(h.value) as type FROM runs as r, jsonb_each(r.hparams) as h WHERE r.hparams is not NULL AND r.hparams !='null') as s WHERE s.type!='object';
+
+INSERT INTO run_hparams(run_id, hparam, number_val, text_val, bool_val)
+SELECT n.id as run_id, n.key as hparam,
+CASE WHEN n.type='number' THEN n.value::text::float ELSE NULL END as number_val,
+CASE WHEN n.type='string' THEN n.value::text ELSE NULL END as text_val,
+CASE WHEN n.type='boolean' THEN n.value::text::boolean ELSE NULL END as bool_val
+FROM (SELECT s.id, CONCAT(s.key, '.', nh.key) as key, nh.value,  jsonb_typeof(nh.value) as type FROM (SELECT r.id, h.key, h.value, jsonb_typeof(h.value) as type FROM runs as r, jsonb_each(r.hparams) as h WHERE r.hparams is not NULL AND r.hparams !='null') as s, jsonb_each(s.value) nh WHERE s.type='object') as n;
```

Comment:
Talked with Piyush about this a while ago. It is rare for customers to have hyperparameters more than 1 depth (i.e hp.learning_rate). This migration does up to depth 2 (i.e hp.param1.param2). For future use though any depth can be added.

Reply:
I would really like to avoid not supporting certain uses. If a customer relies on this feature it would be a bad experience. If it is decided that we don't support a use case then I think we would need to announce that in release notes and let customer success know.

Are we sure there is no way to support any nesting? Can we base this off what @gt2345 did in the past?
https://github.com/determined-ai/determined/blob/31bc08a5df3a1a42974c455039c48a783501de04/master/static/migrations/20230329193551_add-hyperparameters-metrics-name-view.tx.up.sql#L3

Reply:
Added a recursive query.